src/udev/udev-watch.c

```diff
@@ -161,36 +165,450 @@ static int synthesize_change_one(sd_device *dev, sd_device *target) {
+
+        const char *e = getenv("SYSTEMD_UDEV_SYNC_PARTITION_TIMEOUT_SEC");
+        if (!e)
+                return saved;
+
+        usec_t timeout;
+        r = parse_sec(e, &timeout);
+        if (r < 0)
+                log_debug_errno(r, "Failed to parse $SYSTEMD_UDEV_SYNC_PARTITION_TIMEOUT_SEC=%s, ignoring: %m", e);
+
+        if (timeout > 5 * USEC_PER_HOUR) /* Add an arbitrary upper bound for safety */
+                log_debug("Parsed $SYSTEMD_UDEV_SYNC_PARTITION_TIMEOUT_SEC=%s is too large, ignoring.", e);
+        else
+                saved = timeout;
+
+        return saved;
+}
+
+typedef struct SyncDeviceContext {
+        sd_device *dev;
+        sd_event *event;
+        sd_event_source *timer;
+        usec_t delay;
+} SyncDeviceContext;
+
+static void sync_device_context_done(SyncDeviceContext *c) {
+        assert(c);
+
+        sd_device_unref(c->dev);
+        sd_event_unref(c->event);
+        sd_event_source_unref(c->timer);
+}
+
+static int on_watch_timer(sd_event_source *s, uint64_t usec, void *userdata);
+
+static int context_check_partition_synced(SyncDeviceContext *c) {
+        sd_device *dev = ASSERT_PTR(ASSERT_PTR(c)->dev);
+        int r;
+
+        r = partitions_synced(dev);
+        if (r == -EBUSY) {
+                r = event_reset_time_relative(
+                                c->event, &c->timer, CLOCK_MONOTONIC, c->delay, /* accuracy = */ 0,
+                                on_watch_timer, &c, SD_EVENT_PRIORITY_NORMAL,
+                                "watch-timer", /* force_reset = */ false);
+                if (r < 0)
+                        return log_device_debug_errno(dev, r, "Failed to reset timer: %m");
+
+                /* Let's exponentially increase the delay, to make not torture the disk. */
+                c->delay *= 2;
+                return 0;
+        }
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to check if partitions are synced: %m");
+        if (r == 0)
+                c->delay = WAIT_SYNCED_INITIAL_DELAY_USEC;
 
         return r;
 }
 
+static int context_on_event(SyncDeviceContext *c) {
+        int r;
+
+        assert(c);
+
+        r = context_check_partition_synced(c);
+        if (r != 0)
+                return sd_event_exit(c->event, r);
+
+        return 0;
+}
+
+static int on_watch_monitor(sd_device_monitor *monitor, sd_device *dev, void *userdata) {
+        return context_on_event(userdata);
+}
+
+static int on_watch_timer(sd_event_source *s, uint64_t usec, void *userdata) {
+        return context_on_event(userdata);
+}
+
+static int wait_for_partitions_synced(sd_device *dev) {
+        int r;
+
+        assert(dev);
+
+        /* Let's anyway check if partitions are already synced at first. */
+        r = partitions_synced(dev);
+        if (r < 0 && r != -EBUSY)
+                return log_device_debug_errno(dev, r, "Failed to check if partitions are synced: %m");
+        if (r > 0)
+                return r; /* Yay! */
```

Comment:
Isn't this excessive since context_check_partition_synced() will check immediately again later on in this function?